pyro/infer/mcmc/util.py

```diff
@@ -0,0 +1,137 @@
+import torch
+from collections import defaultdict
+
+from pyro.distributions.util import log_sum_exp
+from pyro.infer.util import is_validation_enabled
+from pyro.poutine.indep_messenger import CondIndepStackFrame
+from pyro.util import check_site_shape
+
+
+class EnumTraceProbEvaluator(object):
+    """
+    Computes the log probability density of a trace that possibly contains
+    discrete sample sites enumerated in parallel.
+
+    :param model_trace: execution trace from a static model.
+    :param bool has_enumerable_sites: whether the trace contains any
+        discrete enumerable sites.
+    :param int max_iarange_nesting: Optional bound on max number of nested
+        :func:`pyro.iarange` contexts.
+    """
+    def __init__(self,
+                 model_trace,
+                 has_enumerable_sites=False,
+                 max_iarange_nesting=float("inf")):
+        self.has_enumerable_sites = has_enumerable_sites
+        self.max_iarange_nesting = max_iarange_nesting
+        default_cond_stack = (CondIndepStackFrame(name="default", dim=0, size=0, counter=None),)
+        self.root = frozenset(default_cond_stack)
```

Comment:
Why not the empty `frozenset()`?

Reply:
I think this can be removed now. I was initially doing something like `if frozenset() in frozenset([1])` (which evaluates to False), so this was to not have to put in special cases to deal with the root node. Will remove this now.